From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Arjan van de Ven <arjan@linux.intel.com>
Date: Mon, 13 Feb 2023 14:48:45 +0000
Subject: [PATCH 167/184] net: dst: reduce false sharing in dst_entry

The dst_entry struct has a reference count that gets touched quite a bit
in scenarios where many connections happen from/to the same IP. This
dirty cache line is shared with other members that are read (mostly) that
are used quite a bit.

Add padding to push the refcount to its own cache line, avoiding false
sharing. Benchmarking with memcached (1:100 mode) shows a 30%+ gain.

Signed-off-by: Arjan van de Ven <arjan@linux.intel.com>
Signed-off-by: Wangyang Guo <wangyang.guo@intel.com>
---
 include/net/dst.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/net/dst.h b/include/net/dst.h
index 75e26f0e4537..a1b2c3d4e5f6 100644
--- a/include/net/dst.h
+++ b/include/net/dst.h
@@ -70,6 +70,7 @@ struct dst_entry {
 	 */
 #ifdef CONFIG_64BIT
 	rcuref_t		__rcuref;	/* 64-bit offset 64 */
+	int			__pad2[15];
 #endif
 	int			__use;
 	unsigned long		lastuse;
--
2.47.1

